### ISSUE TITLE
CFE doesn't implicitly instantiate generic method

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/sdk#32654]
In DDC-Analyzer, `f` is implicitly instantiated with `String` when passed to `bar` in `main`:

```
void foo<E>(E s) {
  print(s);
  print(E);
}

void bar(void Function(String) f) {
  f('hi');
}

void main() {
  var f = foo;
  bar(f);
}
```

In CFE, this is a static error:
```
genmethod.dart:13:7: Error: A value of type '<E extends dart.core::Object>(#libnull::foo::E) → void' can't be assigned to a variable of type '(dart.core::String) → void'.
Try changing the type of the left hand side, or casting the right hand side to '(dart.core::String) → void'.
  bar(f);
```

**Comment by vsmenon** [comment]
@lrhn @leafpetersen - what is the correct behavior here?

**Comment by eernstg** [comment]
We have the following feature specification on generic function instantiation in the pipeline: https://dart-review.googlesource.com/c/sdk/+/47043.

If it is landed as-is then your example is covered: A generic function instantiation with type argument `String` must take place in `bar(f)`, so the invocation is statically error-free, the function passed will be "`f<String>`", and it succeeds at run time.

**Comment by vsmenon** [comment]
thanks for the clarification, @eernstg !

**Comment by vsmenon** [comment]
Refiling on language.  If we're changing the analyzer behavior, this is a breaking change - may need to sanity check.

**Comment by lrhn** [comment]
We're not planning on changing the expected behavior at this point, and I believe the CFE is implementing the correct semantics, so both dart2js and the VM is currently running with that behavior in Dart 2 mode.

In which way is this a breaking change? That is, what breaks?

**Comment by fishythefish** [comment]
I'm cleaning up massively broken tests, and this is an issue with `language_2/instantiate_tearoff_of_call_test`.

Here's the important code:
```dart
T f<T>(T x) => x;

void test(T Function<T>(T) f) {
  int Function(int) intFunc = f.call;
}

main() {
  test(f);
}
```

CFE disallows the assignment of `f.call` to `intFunc` because a value of type `T Function(T)` can't be assigned to a variable of type `int Function(int)`.

Analyzer reports no errors, but DDC reports a runtime error.

It sounds like CFE is right, so https://dart-review.googlesource.com/c/sdk/+/116485 updates the test.

Question, though: Why doesn't this fall under the 2nd exception?
* *e* is `f.call`, which is a property extraction denoting a closurization. (`call` is special, perhaps, but this happens even if the generic function type is replaced by a generic class and `call` is replaced by an instance method of type `T Function(T)`.)
* The static type of *e* is the generic function type `T Function<T>(T)`.
* *e* occurs in a context (the RHS of the assignment) where the expected type is the non-generic function type `int Function(int)`.
* Type inference should come up with `T = int`.

**Comment by lrhn** [comment]
This is clever and annoying.
We chose that you can only do generic function instantiation on tear-off, not for an existing function value, but since all function values have `call` members that you can tear off, that's not actually a restriction, just a hurdle. Maybe we should revisit that decision.

I agree that as the specification is written, this is a tear-off of a generic function in a non-generic context, and type inference should cause an instantiation of the torn off function with `int`.

**Comment by eernstg** [comment]
The specification says ([here](https://github.com/dart-lang/language/blob/51c5fd4af78ec6b6edae3862a1bdbf43f160871d/specification/dartLangSpec.tex#L9568)) that

> The expectation is that any concrete subclass of \FUNCTION{} will implement \CALL,

Given that it has always been true that a function object is an instance of a class that implements `Function`, this implies that that every function object implements a method called `call` (otherwise that expectation would definitely not have been reasonable). All other classes are actually irrelevant today, because it is impossible for any other class to implement `Function` (the syntax `implements Function` is not an error, but has no effect).

About the analysis following after
> Question, though..

I agree with the analysis and with @lrhn: `int Function(int) intFunc = f.call;` should not be an error. A function object has a `call` method, and we don't treat it as anything but an instance method anywhere, so we shouldn't have exceptions for it in this case, either.

It does make sense, too: Generic function instantiation is an unusual operation in Dart: It is a semantically significant operation that occurs implicitly based on the static types, and we do not even have an explicit syntax for it (yet). So it's useful if it is likely to be at least somewhat recognizable based on the syntax (that is: at a glance).

The use of `f.call` adds emphasis to the fact that (1) this is a property extraction, and (2) since `call` is recognizable as the name of a method, it looks like a closurization. At that point, generic method instantiation is more likely to come to mind. So it's not so bad after all. ;-)

**Comment by fishythefish** [comment]
Fair enough. :) Leaving the test alone, but currently only analyzer is passing. Filed #38301 and #38302 to track CFE and DDC.

**Comment by fishythefish** [comment]
```dart
void foo<E>(E s) {
  print(s);
  print(E);
}

void bar(void Function(String) f) {
  f('hi');
}

void main() {
  bar(foo);
}
```

The code above is fine. The code below fails to compile (with CFE). What's the distinction?

```dart
void foo<E>(E s) {
  print(s);
  print(E);
}

void bar(void Function(String) f) {
  f('hi');
}

void baz(void Function<E>(E) f) {
  bar(f);
}

void main() {
  baz(foo);
}
```

**Comment by leafpetersen** [comment]
> We chose that you can only do generic function instantiation on tear-off, not for an existing function value, but since all function values have `call` members that you can tear off, that's not actually a restriction, just a hurdle. Maybe we should revisit that decision.

I don't have strong feelings either way, but my recollection is that the reason for the restriction was concerns about implicit changes in identity.  Since tearoffs already are allowed to change identity, the sense was that allowing this on tearoffs was less troublesome.  From that standpoint, an explicit `f.call` arguably marks a point where identity could change, as opposed to just `f`.
